wasm_builder/src/lib.rs

```diff
@@ -0,0 +1,319 @@
+//! Crate with helper tool to build smartcontracts, triggers and permission
+//! validators for Iroha 2.
+//!
+//! See [`Builder`] for more details.
+
+use std::{
+    env,
+    path::{Path, PathBuf},
+    process::Command,
+};
+
+use eyre::{bail, eyre, Context as _, Result};
+
+/// Current toolchain used to build smartcontracts
+const TOOLCHAIN: &str = "+nightly-2022-12-22";
+
+/// WASM Builder for smartcontracts, triggers and permission validators.
+///
+/// # Example
+///
+/// ```no_run
+/// use iroha_wasm_builder::Builder;
+/// use eyre::Result;
+///
+/// fn main() -> Result<()> {
+///     let bytes = Builder::new("relative/path/to/smartcontract/")
+///         .out_dir("path/to/out/dir") // Optional: Set output directory
+///         .format() // Optional: Enable smartcontract formatting
+///         .build()? // Run build
+///         .optimize()? // Optimize WASM output
+///         .into_bytes();
+///
+///     // ...
+///
+///     Ok(())
+/// }
+/// ```
+#[derive(Debug)]
+pub struct Builder {
+    /// Path to the smartcontract or to the directory of smartcontracts
+    path: PathBuf,
+    /// Build output directory
+    out_dir: Option<PathBuf>,
+    /// Flag to enable smartcontract formatting
+    format: bool,
+}
+
+impl Builder {
+    /// Initialize [`Builder`] with path to smartcontract.
+    ///
+    /// `relative_path` should be relative to `CARGO_MANIFEST_DIR`.
+    pub fn new(relative_path: impl Into<PathBuf>) -> Self {
+        Self {
+            path: relative_path.into(),
+            out_dir: None,
```

Comment:
No, because `default_out_dir()` relies on `OUT_DIR` env var, which is defined only in build-scripts and I don't want `new()` to return an error when called from non a build-script.

So not using `out_dir()` is okay in `build.rs` but not in basic code like `kagami` or tests.

I think this crate will be mostly used in build-scripts so *API* is specified for that. Howerver it is also usefull in basic code, you just need to read docs for `build()` function, where it warns you about `out_dir`.

Reply:
I got some likes. I think, I can close this one